File: jsol-compiler-node.bak/1240-python-ternary.js
// @JSOL v0.2.97 - Python Ternary Reorderer
const $bIsIdentChar2 = function($saCh) {
  if ($saCh === "_") {
    return true;
  }
  if ($saCh === "$") {
    return true;
  }
  if ($saCh >= "a" && $saCh <= "z") {
    return true;
  }
  if ($saCh >= "A" && $saCh <= "Z") {
    return true;
  }
  if ($saCh >= "0" && $saCh <= "9") {
    return true;
  }
  return false;
};
const $iFindStatementEnd = function($saCode, $iStart) {
  const $iLen = Str["len"]($saCode);
    let $iParenDepth = 0;
    let $iBracketDepth = 0;
    let $i = $iStart;

    while ($i < $iLen) {
    const $saCh = Str["sub"]($saCode,  $i,  1);
        if ($saCh === "(") {
      $iParenDepth = $iParenDepth + 1;
    }
    else if ($saCh === ")") {
      $iParenDepth = $iParenDepth - 1;
    }
    else if ($saCh === "[") {
      $iBracketDepth = $iBracketDepth + 1;
    }
    else if ($saCh === "]") {
      $iBracketDepth = $iBracketDepth - 1;
    }
    else if ($saCh === ";" && $iParenDepth === 0 && $iBracketDepth === 0) {
      return $i;
    }
    $i = $i + 1;
  }
  return $iLen;
};
const $mSplitTernary = function($saExpr) {
  const $iLen = Str["len"]($saExpr);
    let $iParenDepth = 0;
    let $iBracketDepth = 0;
    let $iQuestionIndex = -1;
    let $iColonIndex = -1;
    let $i = 0;

    while ($i < $iLen) {
    const $saCh = Str["sub"]($saExpr,  $i,  1);
        if ($saCh === "(") {
      $iParenDepth = $iParenDepth + 1;
    }
    else if ($saCh === ")") {
      $iParenDepth = $iParenDepth - 1;
    }
    else if ($saCh === "[") {
      $iBracketDepth = $iBracketDepth + 1;
    }
    else if ($saCh === "]") {
      $iBracketDepth = $iBracketDepth - 1;
    }
    else if ($saCh === "?" && $iParenDepth === 0 && $iBracketDepth === 0) {
      if ($iQuestionIndex === -1) {
        $iQuestionIndex = $i;
      }
      else {
        return JSOL.dict("ok",  false);
      }
    }
    else if ($saCh === ":" && $iParenDepth === 0 && $iBracketDepth === 0 && $iQuestionIndex !== -1 && $iColonIndex === -1) {
      $iColonIndex = $i;
    }
    $i = $i + 1;
  }
  if ($iQuestionIndex === -1 || $iColonIndex === -1) {
    return JSOL.dict("ok",  false);
  }
  const $saCond = $saTrimWhitespace(Str["sub"]($saExpr,  0,  $iQuestionIndex));
    const $saTrue = $saTrimWhitespace(Str["sub"]($saExpr,  $iQuestionIndex + 1,  $iColonIndex - ($iQuestionIndex + 1)));
    const $saFalse = $saTrimWhitespace(Str["sub"]($saExpr,  $iColonIndex + 1,  $iLen - ($iColonIndex + 1)));

    return JSOL.dict("ok",  true,  "cond",  $saCond,  "true",  $saTrue,  "false",  $saFalse);
};
const $saConvertTernaries = function($saMaskedCode) {
  let $saResult = "";
    let $i = 0;
    const $iLen = Str["len"]($saMaskedCode);

    while ($i < $iLen) {
    const $saCh = Str["sub"]($saMaskedCode,  $i,  1);
        const $bAtBoundary = ($i === 0) || ($bIsIdentChar2(Str["sub"]($saMaskedCode,  $i - 1,  1)) === false);
        let $bHandled = false;

        if ($bAtBoundary === true && Str["sub"]($saMaskedCode,  $i,  7) === "return " && $bIsIdentChar2(Str["sub"]($saMaskedCode,  $i + 6,  1)) === false) {
      const $iRhsStart = $i + 7;
            const $iStmtEnd = $iFindStatementEnd($saMaskedCode, $iRhsStart);
            const $saRhs = Str["sub"]($saMaskedCode,  $iRhsStart,  $iStmtEnd - $iRhsStart);
            const $mSplit = $mSplitTernary($saRhs);

            if ($mSplit["ok"] === true) {
        $saResult = $saResult + "" + "return (" + "" + $mSplit["true"] + "" + " if " + "" + $mSplit["cond"] + "" + " else " + "" + $mSplit["false"] + "" + ")" + ";";
                $i = $iStmtEnd + 1;
                $bHandled = true;
      }
    }
    if ($bHandled === false && $saCh === "=") {
      const $saPrevCh = Str["sub"]($saMaskedCode,  $i - 1,  1);
            const $saNextCh = Str["sub"]($saMaskedCode,  $i + 1,  1);
            const $bIsPlainAssign = (($saNextCh !== "=") && ($saPrevCh !== "=") && ($saPrevCh !== "<") && ($saPrevCh !== ">") && ($saPrevCh !== "!"));

            if ($bIsPlainAssign === true) {
        const $iRhsStart = $i + 1;
                const $iStmtEnd = $iFindStatementEnd($saMaskedCode, $iRhsStart);
                const $saRhs = Str["sub"]($saMaskedCode,  $iRhsStart,  $iStmtEnd - $iRhsStart);
                const $mSplit = $mSplitTernary($saRhs);

                if ($mSplit["ok"] === true) {
          $saResult = $saResult + "" + "= (" + "" + $mSplit["true"] + "" + " if " + "" + $mSplit["cond"] + "" + " else " + "" + $mSplit["false"] + "" + ")" + ";";
                    $i = $iStmtEnd + 1;
                    $bHandled = true;
        }
      }
    }
    if ($bHandled === false) {
      $saResult = $saResult + "" + $saCh;
            $i = $i + 1;
    }
  }
  return $saResult;
};

File: jsol-compiler-py.bak/1240-python-ternary.py
import math
import functools
from jsol_core import JSOL

# @JSOL v0.2.97 - Python Ternary Reorderer
def bIsIdentChar2(saCh): 

  if saCh == "_": 

    return True;


  if saCh == "$": 

    return True;


  if saCh >= "a" and saCh <= "z": 

    return True;


  if saCh >= "A" and saCh <= "Z": 

    return True;


  if saCh >= "0" and saCh <= "9": 

    return True;


  return False;


def iFindStatementEnd(saCode, iStart): 

  iLen = len(saCode);
  iParenDepth = 0;
  iBracketDepth = 0;
  i = iStart;

  while i < iLen: 

    saCh = saCode[( i):( i)+( 1)];
    if saCh == "(": 

      iParenDepth = iParenDepth + 1;


    elif saCh == ")": 

      iParenDepth = iParenDepth - 1;


    elif saCh == "[": 

      iBracketDepth = iBracketDepth + 1;


    elif saCh == "]": 

      iBracketDepth = iBracketDepth - 1;


    elif saCh == ";" and iParenDepth == 0 and iBracketDepth == 0: 

      return i;


    i = i + 1;


  return iLen;


def mSplitTernary(saExpr): 

  iLen = len(saExpr);
  iParenDepth = 0;
  iBracketDepth = 0;
  iQuestionIndex = -1;
  iColonIndex = -1;
  i = 0;

  while i < iLen: 

    saCh = saExpr[( i):( i)+( 1)];
    if saCh == "(": 

      iParenDepth = iParenDepth + 1;


    elif saCh == ")": 

      iParenDepth = iParenDepth - 1;


    elif saCh == "[": 

      iBracketDepth = iBracketDepth + 1;


    elif saCh == "]": 

      iBracketDepth = iBracketDepth - 1;


    elif saCh == "?" and iParenDepth == 0 and iBracketDepth == 0: 

      if iQuestionIndex == -1: 

        iQuestionIndex = i;


      else: 

        return JSOL.dict("ok",  False);




    elif saCh == ":" and iParenDepth == 0 and iBracketDepth == 0 and iQuestionIndex != -1 and iColonIndex == -1: 

      iColonIndex = i;


    i = i + 1;


  if iQuestionIndex == -1 or iColonIndex == -1: 

    return JSOL.dict("ok",  False);


  saCond = saTrimWhitespace(saExpr[( 0):( 0)+( iQuestionIndex)]);
  saTrue = saTrimWhitespace(saExpr[( iQuestionIndex + 1):( iQuestionIndex + 1)+( iColonIndex - (iQuestionIndex + 1))]);
  saFalse = saTrimWhitespace(saExpr[( iColonIndex + 1):( iColonIndex + 1)+( iLen - (iColonIndex + 1))]);

  return JSOL.dict("ok",  True,  "cond",  saCond,  "true",  saTrue,  "false",  saFalse);


def saConvertTernaries(saMaskedCode): 

  saResult = "";
  i = 0;
  iLen = len(saMaskedCode);

  while i < iLen: 

    saCh = saMaskedCode[( i):( i)+( 1)];
    bAtBoundary = (i == 0) or (bIsIdentChar2(saMaskedCode[( i - 1):( i - 1)+( 1)]) == False);
    bHandled = False;

    if bAtBoundary == True and saMaskedCode[( i):( i)+( 7)] == "return " and bIsIdentChar2(saMaskedCode[( i + 6):( i + 6)+( 1)]) == False: 

      iRhsStart = i + 7;
      iStmtEnd = iFindStatementEnd(saMaskedCode, iRhsStart);
      saRhs = saMaskedCode[( iRhsStart):( iRhsStart)+( iStmtEnd - iRhsStart)];
      mSplit = mSplitTernary(saRhs);

      if mSplit["ok"] == True: 

        saResult = saResult + "" + "return (" + "" + mSplit["true"] + "" + " if " + "" + mSplit["cond"] + "" + " else " + "" + mSplit["false"] + "" + ")" + ";";
        i = iStmtEnd + 1;
        bHandled = True;




    if bHandled == False and saCh == "=": 

      saPrevCh = saMaskedCode[( i - 1):( i - 1)+( 1)];
      saNextCh = saMaskedCode[( i + 1):( i + 1)+( 1)];
      bIsPlainAssign = ((saNextCh != "=") and (saPrevCh != "=") and (saPrevCh != "<") and (saPrevCh != ">") and (saPrevCh != "!"));

      if bIsPlainAssign == True: 

        iRhsStart = i + 1;
        iStmtEnd = iFindStatementEnd(saMaskedCode, iRhsStart);
        saRhs = saMaskedCode[( iRhsStart):( iRhsStart)+( iStmtEnd - iRhsStart)];
        mSplit = mSplitTernary(saRhs);

        if mSplit["ok"] == True: 

          saResult = saResult + "" + "= (" + "" + mSplit["true"] + "" + " if " + "" + mSplit["cond"] + "" + " else " + "" + mSplit["false"] + "" + ")" + ";";
          i = iStmtEnd + 1;
          bHandled = True;






    if bHandled == False: 

      saResult = saResult + "" + saCh;
      i = i + 1;




  return saResult;


